main.cpp

```diff
@@ -108,7 +108,410 @@ write 3 UDTs below that EACH have:
+    ~Loan();
+    void calculatePayments();
+    double calculateOverpayment();
+};
+//--
+Loan::Loan(double amount, double rate, unsigned long numPay, bool secured):
+    loanAmount(amount), interestRateInPercent(rate), numPayments(numPay), isSecured(secured)
+{
+    if(isSecured)
+    {
+        interestRateInPercent *= 0.9;
+    }
+    calculatePayments();
+}
+//--
+Loan::~Loan()
+{
+    std::cout << "Loan destructor called!" << std::endl;
+}
+//--
+void Loan::calculatePayments()
+{
+    payments.clear();
+    for(unsigned int i = 0; i < numPayments; i++)
+        {
+            double currentLoanAmount = loanAmount;
+            double payment = currentLoanAmount * (interestRateInPercent / 12 / 100.0) + loanAmount / static_cast<double>(numPayments);
+            currentLoanAmount -= loanAmount / static_cast<double>(numPayments);
+            payments.push_back(payment);
+        }
+}
+//--
+double Loan::calculateOverpayment()
+{
+    double amountToPay = 0.0;
+    for(unsigned int i = 0; i < numPayments; i++)
+        {
+            amountToPay += payments[i];
+        }
+    return amountToPay - loanAmount;
+}
+//--
+struct Gym 
+{
+    struct Member 
+    {
+        std::string name;
+        static int memberId;
+        int age;
+        bool isActive = true;
+        std::vector<std::string> activityLog;
+
+        Member(std::string n, int a);
+        ~Member();
+        void logActivity(const std::string& activity);
+        void printDetails();
+    };
+
+    std::vector<Member> members;
+    std::string gymName;
+    int totalMembers = 0;
+    double membershipPrice = 30.0;
+    double monthlyRevenue = 0.0;
+    bool isOpen;
+
+    Gym(std::string name, bool open);
+    ~Gym();
+    void registerMember(const Member& member);
+    void toggleGymStatus();
+    void calculateRevenue();
+};
+//--
+int Gym::Member::memberId = 0;
+//--
+Gym::Member::Member(std::string n, int a):
+    name(n), age(a)
+{
+    memberId++;
+}
+//--
+Gym::Member::~Member()
+{
+    std::cout << "Member destructor called!" << std::endl;
+}
+//--
+void Gym::Member::logActivity(const std::string& activity) 
+{
+    activityLog.push_back(activity);
+}
+//--
+void Gym::Member::printDetails()
+{
+    std::cout << "Member ID: " << memberId << ", Name: " << name << ", Age: " << age << ", Active: " << (isActive ? "Yes": "No") << std::endl;
+    std::cout << "Activities Logged: ";
+    for (const auto& act : activityLog) 
+    {
+        std::cout << act << " ";
+    }
+    std::cout << std::endl;
+}
+Gym::Gym(std::string name, bool open) : 
+    gymName(name), isOpen(open)
+{
+}
+//--
+Gym::~Gym()
+{
+    std::cout << "Gym destructor called!" << std::endl;
+}
+//--
+void Gym::registerMember(const Member& member) 
+{
+    members.push_back(member);
+    totalMembers++;
+}
+//--
+void Gym::toggleGymStatus() 
+{
+    isOpen = !isOpen;
+}
+//--
+void Gym::calculateRevenue() 
+{
+    monthlyRevenue = totalMembers * membershipPrice;
+}
+//--
+struct Classroom 
+{
+public:
```

Comment:
`struct` is public by default. This keyword is unnecessary and hasn't been taught yet in the course. 
- [ ] revise